{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exercises - week 2\n",
    "\n",
    "Write your solutions to the given cells (see \"WRITE YOUR CODE HERE\"). Make your code print only what is requested. Select `Kernel` &rightarrow; `Restart & Clear Output` and then `Cell` &rightarrow; `Run All` from the menu before submitting your solutions, so that your submission contains clean output produced from scratch."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.1\n",
    "\n",
    "Print the ids of those entries that match the protein name \"p53\" and the organism \"Homo sapiens\" in the NCBI Protein database, one per line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio.Entrez as BE\n",
    "import Bio.SeqIO as BSIO\n",
    "# set email globally\n",
    "BE.email = 'jojuket@utu.fi'\n",
    "# remember to set your real email address\n",
    "# before continuing to the examples below\n",
    "#SET_YOUR_REAL_EMAIL_ADDRESS_BEFORE_COMMENTING_OUT_THIS_LINE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID's are:\n",
      "454545707\n",
      "454545207\n",
      "454544715\n",
      "454544286\n",
      "454543597\n",
      "454521557\n",
      "454520873\n",
      "371502119\n",
      "187830909\n",
      "187830901\n",
      "187830894\n",
      "187830855\n",
      "187830823\n",
      "187830777\n",
      "120407068\n",
      "269849759\n",
      "232816\n",
      "232814\n",
      "189476\n",
      "349734072\n"
     ]
    }
   ],
   "source": [
    "handle = BE.esearch(db=\"protein\", term=\"Homo sapiens[Organism] p53[Gene Name]\")\n",
    "# get response from NCBI\n",
    "# (data as bytes, not as string)\n",
    "results = handle.read()\n",
    "# close the handle\n",
    "handle.close()\n",
    "# save results to file\n",
    "# ('wb' needed to write bytes)\n",
    "with open('results.xml', 'wb') as f:\n",
    "    f.write(results)\n",
    "    # load from file and parse\n",
    "# ('rb' needed to read data as bytes)\n",
    "with open('results.xml', 'rb') as f:\n",
    "    results = BE.read(f)\n",
    "# report the retrieved IDs\n",
    "# print(results['IdList'])\n",
    "    \n",
    "    # Probably should've done this somehow else but didn't know how so my style was:\n",
    "    # Extract the IdList from the results\n",
    "    pIDS = results['IdList']\n",
    "    print(\"ID's are:\")\n",
    "    #make a for loop to print the ID's one by one\n",
    "    for i in pIDS:\n",
    "        print(i)\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.2\n",
    "\n",
    "Fetch the entry NP_004955 from the NCBI Protein database and save it to a file named `NP_004955.gb`. Use the GenPept flat file format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NP_004955                482 aa            linear   PRI 18-JAN-2021\n",
      "DEFINITION  histone deacetylase 1 [Homo sapiens].\n",
      "ACCESSION   NP_004955\n",
      "VERSION     NP_004955.2\n",
      "DBSOURCE    REFSEQ: accession NM_004964.3\n",
      "KEYWORDS    RefSeq; MANE Select.\n",
      "SOURCE      Homo sapiens (human)\n",
      "  ORGANISM  Homo sapiens\n",
      "            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\n",
      "            Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;\n",
      "            Catarrhini; Hominidae; Homo.\n",
      "REFERENCE   1  (residues 1 to 482)\n",
      "  AUTHORS   Lv L, Zhang D, Hua P and Yang S.\n",
      "  TITLE     The glial-specific hypermethylated 3' untranslated region of\n",
      "            histone deacetylase 1 may modulates several signal pathways in\n",
      "            Alzheimer's disease\n",
      "  JOURNAL   Life Sci 265, 118760 (2021)\n",
      "   PUBMED   33212149\n",
      "  REMARK    GeneRIF: The glial-specific hypermethylated 3' untranslated region\n",
      "            of histone deacetylase 1 may modulates several signal pathways in\n",
      "            Alzheimer's disease.\n",
      "REFERENCE   2  (residues 1 to 482)\n",
      "  AUTHORS   Bastola S, Pavlyukov MS, Yamashita D, Ghosh S, Cho H, Kagaya N,\n",
      "            Zhang Z, Minata M, Lee Y, Sadahiro H, Yamaguchi S, Komarova S, Yang\n",
      "            E, Markert J, Nabors LB, Bhat K, Lee J, Chen Q, Crossman DK,\n",
      "            Shin-Ya K, Nam DH and Nakano I.\n",
      "  TITLE     Glioma-initiating cells at tumor edge gain signals from tumor core\n",
      "            cells to promote their malignancy\n",
      "  JOURNAL   Nat Commun 11 (1), 4660 (2020)\n",
      "   PUBMED   32938908\n",
      "  REMARK    GeneRIF: Glioma-initiating cells at tumor edge gain signals from\n",
      "            tumor core cells to promote their malignancy.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   3  (residues 1 to 482)\n",
      "  AUTHORS   Shao LW, Peng Q, Dong M, Gao K, Li Y, Li Y, Li CY and Liu Y.\n",
      "  TITLE     Histone deacetylase HDA-1 modulates mitochondrial stress response\n",
      "            and longevity\n",
      "  JOURNAL   Nat Commun 11 (1), 4639 (2020)\n",
      "   PUBMED   32934238\n",
      "  REMARK    GeneRIF: Histone deacetylase HDA-1 modulates mitochondrial stress\n",
      "            response and longevity.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   4  (residues 1 to 482)\n",
      "  AUTHORS   Haenig C, Atias N, Taylor AK, Mazza A, Schaefer MH, Russ J,\n",
      "            Riechers SP, Jain S, Coughlin M, Fontaine JF, Freibaum BD,\n",
      "            Brusendorf L, Zenkner M, Porras P, Stroedicke M, Schnoegl S,\n",
      "            Arnsburg K, Boeddrich A, Pigazzini L, Heutink P, Taylor JP,\n",
      "            Kirstein J, Andrade-Navarro MA, Sharan R and Wanker EE.\n",
      "  TITLE     Interactome Mapping Provides a Network of Neurodegenerative Disease\n",
      "            Proteins and Uncovers Widespread Protein Aggregation in Affected\n",
      "            Brains\n",
      "  JOURNAL   Cell Rep 32 (7), 108050 (2020)\n",
      "   PUBMED   32814053\n",
      "REFERENCE   5  (residues 1 to 482)\n",
      "  AUTHORS   Sixto-Lopez Y, Gomez-Vidal JA, de Pedro N, Bello M,\n",
      "            Rosales-Hernandez MC and Correa-Basurto J.\n",
      "  TITLE     Hydroxamic acid derivatives as HDAC1, HDAC6 and HDAC8 inhibitors\n",
      "            with antiproliferative activity in cancer cell lines\n",
      "  JOURNAL   Sci Rep 10 (1), 10462 (2020)\n",
      "   PUBMED   32591593\n",
      "  REMARK    GeneRIF: Hydroxamic acid derivatives as HDAC1, HDAC6 and HDAC8\n",
      "            inhibitors with antiproliferative activity in cancer cell lines.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   6  (residues 1 to 482)\n",
      "  AUTHORS   Taunton J, Hassig CA and Schreiber SL.\n",
      "  TITLE     A mammalian histone deacetylase related to the yeast\n",
      "            transcriptional regulator Rpd3p\n",
      "  JOURNAL   Science 272 (5260), 408-411 (1996)\n",
      "   PUBMED   8602529\n",
      "REFERENCE   7  (residues 1 to 482)\n",
      "  AUTHORS   Wolffe AP.\n",
      "  TITLE     Histone deacetylase: a regulator of transcription\n",
      "  JOURNAL   Science 272 (5260), 371-372 (1996)\n",
      "   PUBMED   8602525\n",
      "  REMARK    Review article\n",
      "REFERENCE   8  (residues 1 to 482)\n",
      "  AUTHORS   Furukawa Y, Kawakami T, Sudo K, Inazawa J, Matsumine A, Akiyama T\n",
      "            and Nakamura Y.\n",
      "  TITLE     Isolation and mapping of a human gene (RPD3L1) that is homologous\n",
      "            to RPD3, a transcription factor in Saccharomyces cerevisiae\n",
      "  JOURNAL   Cytogenet Cell Genet 73 (1-2), 130-133 (1996)\n",
      "   PUBMED   8646880\n",
      "REFERENCE   9  (residues 1 to 482)\n",
      "  AUTHORS   Hijmans EM, Voorhoeve PM, Beijersbergen RL, van 't Veer LJ and\n",
      "            Bernards R.\n",
      "  TITLE     E2F-5, a new E2F family member that interacts with p130 in vivo\n",
      "  JOURNAL   Mol Cell Biol 15 (6), 3082-3089 (1995)\n",
      "   PUBMED   7760804\n",
      "REFERENCE   10 (residues 1 to 482)\n",
      "  AUTHORS   Margolis DM, Somasundaran M and Green MR.\n",
      "  TITLE     Human transcription factor YY1 represses human immunodeficiency\n",
      "            virus type 1 transcription and virion production\n",
      "  JOURNAL   J Virol 68 (2), 905-910 (1994)\n",
      "   PUBMED   8289393\n",
      "COMMENT     REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "            reference sequence was derived from AL109945.15.\n",
      "            This sequence is a reference standard in the RefSeqGene project.\n",
      "            On Feb 26, 2001 this sequence version replaced NP_004955.1.\n",
      "            \n",
      "            Summary: Histone acetylation and deacetylation, catalyzed by\n",
      "            multisubunit complexes, play a key role in the regulation of\n",
      "            eukaryotic gene expression. The protein encoded by this gene\n",
      "            belongs to the histone deacetylase/acuc/apha family and is a\n",
      "            component of the histone deacetylase complex. It also interacts\n",
      "            with retinoblastoma tumor-suppressor protein and this complex is a\n",
      "            key element in the control of cell proliferation and\n",
      "            differentiation. Together with metastasis-associated protein-2, it\n",
      "            deacetylates p53 and modulates its effect on cell growth and\n",
      "            apoptosis. [provided by RefSeq, Jul 2008].\n",
      "            \n",
      "            Publication Note:  This RefSeq record includes a subset of the\n",
      "            publications that are available for this gene. Please see the Gene\n",
      "            record to access additional publications.\n",
      "            \n",
      "            ##Evidence-Data-START##\n",
      "            Transcript exon combination :: SRR1163657.10339.1,\n",
      "                                           SRR1803611.177301.1 [ECO:0000332]\n",
      "            RNAseq introns              :: single sample supports all introns\n",
      "                                           SAMEA1965299, SAMEA1966682\n",
      "                                           [ECO:0000348]\n",
      "            ##Evidence-Data-END##\n",
      "            \n",
      "            ##RefSeq-Attributes-START##\n",
      "            MANE Ensembl match     :: ENST00000373548.8/ ENSP00000362649.3\n",
      "            RefSeq Select criteria :: based on conservation, expression,\n",
      "                                      longest protein\n",
      "            ##RefSeq-Attributes-END##\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..482\n",
      "                     /organism=\"Homo sapiens\"\n",
      "                     /db_xref=\"taxon:9606\"\n",
      "                     /chromosome=\"1\"\n",
      "                     /map=\"1p35.2-p35.1\"\n",
      "     Protein         1..482\n",
      "                     /product=\"histone deacetylase 1\"\n",
      "                     /EC_number=\"3.5.1.98\"\n",
      "                     /note=\"reduced potassium dependency, yeast homolog-like 1\"\n",
      "                     /calculated_mol_wt=54972\n",
      "     Region          4..374\n",
      "                     /region_name=\"HDAC1\"\n",
      "                     /note=\"Histone deacetylase 1 (HDAC1); cd10010\"\n",
      "                     /db_xref=\"CDD:212534\"\n",
      "     Region          9..321\n",
      "                     /region_name=\"Histone deacetylase\"\n",
      "                     /note=\"propagated from UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            74\n",
      "                     /site_type=\"acetylation\"\n",
      "                     /note=\"N6-acetyllysine, alternate.\n",
      "                     /evidence=ECO:0000244|PubMed:19608861; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            order(140..141,149..150,176,178,264,271,301,303)\n",
      "                     /site_type=\"active\"\n",
      "                     /db_xref=\"CDD:212534\"\n",
      "     Site            order(176,178,264)\n",
      "                     /site_type=\"other\"\n",
      "                     /note=\"Zn binding site [ion binding]\"\n",
      "                     /db_xref=\"CDD:212534\"\n",
      "     Site            220\n",
      "                     /site_type=\"acetylation\"\n",
      "                     /note=\"N6-acetyllysine.\n",
      "                     /evidence=ECO:0000244|PubMed:19608861; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            393\n",
      "                     /site_type=\"phosphorylation\"\n",
      "                     /note=\"Phosphoserine.\n",
      "                     /evidence=ECO:0000244|PubMed:17081983,\n",
      "                     ECO:0000244|PubMed:17487921, ECO:0000244|PubMed:18691976,\n",
      "                     ECO:0000244|PubMed:19690332, ECO:0000244|PubMed:20068231,\n",
      "                     ECO:0000244|PubMed:21406692, ECO:0000244|PubMed:23186163,\n",
      "                     ECO:0000244|PubMed:24275569; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            406\n",
      "                     /site_type=\"phosphorylation\"\n",
      "                     /note=\"Phosphoserine.\n",
      "                     /evidence=ECO:0000250|UniProtKB:Q92769; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            409\n",
      "                     /site_type=\"phosphorylation\"\n",
      "                     /note=\"Phosphoserine.\n",
      "                     /evidence=ECO:0000244|PubMed:23186163; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            421\n",
      "                     /site_type=\"phosphorylation\"\n",
      "                     /note=\"Phosphoserine, by CK2.\n",
      "                     /evidence=ECO:0000244|PubMed:17081983,\n",
      "                     ECO:0000244|PubMed:18318008, ECO:0000244|PubMed:19690332,\n",
      "                     ECO:0000244|PubMed:20068231, ECO:0000244|PubMed:21406692,\n",
      "                     ECO:0000244|PubMed:23186163, ECO:0000269|PubMed:11602581;\n",
      "                     propagated from UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            423\n",
      "                     /site_type=\"phosphorylation\"\n",
      "                     /note=\"Phosphoserine, by CK2.\n",
      "                     /evidence=ECO:0000244|PubMed:17081983,\n",
      "                     ECO:0000244|PubMed:18318008, ECO:0000244|PubMed:20068231,\n",
      "                     ECO:0000244|PubMed:21406692, ECO:0000244|PubMed:23186163,\n",
      "                     ECO:0000269|PubMed:11602581; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     Site            432\n",
      "                     /site_type=\"methylation\"\n",
      "                     /note=\"N6-methylated lysine, by EHMT2.\n",
      "                     /evidence=ECO:0000269|PubMed:18438403; propagated from\n",
      "                     UniProtKB/Swiss-Prot (Q13547.1)\"\n",
      "     CDS             1..482\n",
      "                     /gene=\"HDAC1\"\n",
      "                     /gene_synonym=\"GON-10; HD1; KDAC1; RPD3; RPD3L1\"\n",
      "                     /coded_by=\"NM_004964.3:88..1536\"\n",
      "                     /db_xref=\"CCDS:CCDS360.1\"\n",
      "                     /db_xref=\"GeneID:3065\"\n",
      "                     /db_xref=\"HGNC:HGNC:4852\"\n",
      "                     /db_xref=\"MIM:601241\"\n",
      "ORIGIN      \n",
      "        1 maqtqgtrrk vcyyydgdvg nyyygqghpm kphrirmthn lllnyglyrk meiyrphkan\n",
      "       61 aeemtkyhsd dyikflrsir pdnmseyskq mqrfnvgedc pvfdglfefc qlstggsvas\n",
      "      121 avklnkqqtd iavnwagglh hakkseasgf cyvndivlai lellkyhqrv lyididihhg\n",
      "      181 dgveeafytt drvmtvsfhk ygeyfpgtgd lrdigagkgk yyavnyplrd giddesyeai\n",
      "      241 fkpvmskvme mfqpsavvlq cgsdslsgdr lgcfnltikg hakcvefvks fnlpmlmlgg\n",
      "      301 ggytirnvar cwtyetaval dteipnelpy ndyfeyfgpd fklhispsnm tnqntneyle\n",
      "      361 kikqrlfenl rmlphapgvq mqaipedaip eesgdededd pdkrisicss dkriaceeef\n",
      "      421 sdseeegegg rknssnfkka krvktedeke kdpeekkevt eeektkeekp eakgvkeevk\n",
      "      481 la\n",
      "//\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# fetch records from \"protein\" database\n",
    "# and return sequences in GenBank flat file format\n",
    "handle = BE.efetch(db=\"protein\",\n",
    "                   id=\"NP_004955\",\n",
    "                   rettype=\"gb\",\n",
    "                   retmode=\"text\")\n",
    "# get response from NCBI\n",
    "# (data as string, not bytes)\n",
    "results = handle.read()\n",
    "# close the handle\n",
    "handle.close()\n",
    "# save to file\n",
    "# ('w' needed to write string)\n",
    "with open('NP_004955.gb', 'w') as f:\n",
    "    f.write(results)\n",
    "# show raw file content to check that the download was a succé \n",
    "with open('NP_004955.gb') as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.3\n",
    "\n",
    "Load the file `NP_004955.gb` and print the following pieces of information about the protein, one per line.\n",
    "- id\n",
    "- name\n",
    "- description\n",
    "- organism\n",
    "- number of features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ID: NP_004955.2\n",
      " Name: NP_004955\n",
      " Description: histone deacetylase 1 [Homo sapiens]\n",
      " Organism: Homo sapiens\n",
      " Number of features: 15\n"
     ]
    }
   ],
   "source": [
    "# parse entries in GenBank flat file format\n",
    "p53Rec = BSIO.parse('NP_004955.gb', 'gb')\n",
    "#using string formatting %s and %i combined to a for loop to get the required info\n",
    "\n",
    "for f in p53Rec:\n",
    "    print(\" ID: %s\\n Name: %s\\n Description: %s\\n Organism: %s\\n Number of features: %i\" \n",
    "          % (f.id, f.name, f.description, f.annotations['organism'], len(f.features)))  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.4\n",
    "\n",
    "Print the number of phosphorylation sites annotated in the NCBI Protein database entry NP_004955."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phosphorylated sites: 5\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "i = []\n",
    "for rec in BSIO.parse('NP_004955.gb', 'gb'):\n",
    "    #make rec to loop over the NP_004955.gb file\n",
    "    for feature in rec.features:\n",
    "        #move into features\n",
    "        if feature.type == \"Site\":\n",
    "            # get only values from keyword \"Site\"\n",
    "            i= i+feature.qualifiers[\"site_type\"]\n",
    "            #append the values from the \"site_type\" to a list i\n",
    "\n",
    "#initalize the counter with the list i            \n",
    "c = Counter(i)\n",
    "#count the occurences of word phosphorylation\n",
    "print(\"Phosphorylated sites: \"+ str(c['phosphorylation']))\n",
    "            \n",
    "# This was a bit of a turn-a-round, but basically it seems I could do it like this"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.5\n",
    "\n",
    "Regarding the NCBI Protein database entry NP_004955, print the phosphorylated amino acid residues and their positions in the sequence, one per line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phosphorylation[392:393]\n",
      "['Phosphoserine. /evidence=ECO:0000244|PubMed:17081983, ECO:0000244|PubMed:17487921, ECO:0000244|PubMed:18691976, ECO:0000244|PubMed:19690332, ECO:0000244|PubMed:20068231, ECO:0000244|PubMed:21406692, ECO:0000244|PubMed:23186163, ECO:0000244|PubMed:24275569; propagated from UniProtKB/Swiss-Prot (Q13547.1)']\n",
      "phosphorylation[405:406]\n",
      "['Phosphoserine. /evidence=ECO:0000250|UniProtKB:Q92769; propagated from UniProtKB/Swiss-Prot (Q13547.1)']\n",
      "phosphorylation[408:409]\n",
      "['Phosphoserine. /evidence=ECO:0000244|PubMed:23186163; propagated from UniProtKB/Swiss-Prot (Q13547.1)']\n",
      "phosphorylation[420:421]\n",
      "['Phosphoserine, by CK2. /evidence=ECO:0000244|PubMed:17081983, ECO:0000244|PubMed:18318008, ECO:0000244|PubMed:19690332, ECO:0000244|PubMed:20068231, ECO:0000244|PubMed:21406692, ECO:0000244|PubMed:23186163, ECO:0000269|PubMed:11602581; propagated from UniProtKB/Swiss-Prot (Q13547.1)']\n",
      "phosphorylation[422:423]\n",
      "['Phosphoserine, by CK2. /evidence=ECO:0000244|PubMed:17081983, ECO:0000244|PubMed:18318008, ECO:0000244|PubMed:20068231, ECO:0000244|PubMed:21406692, ECO:0000244|PubMed:23186163, ECO:0000269|PubMed:11602581; propagated from UniProtKB/Swiss-Prot (Q13547.1)']\n"
     ]
    }
   ],
   "source": [
    "#Actually couldn't be more proud of myself, I did it, I solved the problem!!!!!!!!\n",
    "#Maybe too early to say, but I feel that i actually learnt something!!!\n",
    "\n",
    "# Make an SeqFeature? object\n",
    "for rec in BSIO.parse('NP_004955.gb', 'gb'):\n",
    "    # intialize variable feature to go through the features of the Seq object\n",
    "    for feature in rec.features:\n",
    "        #Determine that the we are looking for the word \"Site\"\n",
    "        if feature.type == \"Site\":\n",
    "            # initialize variable x to go through site_types and get the values\n",
    "            for x in feature.qualifiers.get(\"site_type\"):\n",
    "                # the key word is \"phosphorylation\"\n",
    "                if x == \"phosphorylation\":\n",
    "                    #then print the site_type x and the location of it\n",
    "                    print(x + str(feature.location))\n",
    "                    print(feature.qualifiers.get(\"note\"))\n",
    "                    \n",
    "#If I print the name of the phosphorylated residue as in print(feature.qualifiers.get(\"note\")), \n",
    "#the Jupyter prints out both /note and /evidence parts. \n",
    "#Could this be because the NCBI data doesn't have \"\" around the name of the amino acid residue (/note part)?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.6\n",
    "\n",
    "The NCBI Nucleotide database entry 223468685 contains several genes. Print their names in the order of appearance, one per line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'BE' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-73025558e99b>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# fetch records from \"nucleotide\" database\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;31m# and return sequences in GenBank flat file format\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m handle = BE.efetch(db=\"nucleotide\",\n\u001b[0m\u001b[0;32m      4\u001b[0m                    \u001b[0mid\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"223468685\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m                    \u001b[0mrettype\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"gb\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'BE' is not defined"
     ]
    }
   ],
   "source": [
    "# fetch records from \"nucleotide\" database\n",
    "# and return sequences in GenBank flat file format\n",
    "handle = BE.efetch(db=\"nucleotide\",\n",
    "                   id=\"223468685\",\n",
    "                   rettype=\"gb\",\n",
    "                   retmode=\"text\")\n",
    "# get response from NCBI\n",
    "# (data as string, not bytes)\n",
    "results = handle.read()\n",
    "# close the handle\n",
    "handle.close()\n",
    "# save to file\n",
    "# ('w' needed to write string)\n",
    "with open('geneEntry.gb', 'w') as f:\n",
    "    f.write(results)\n",
    "    \n",
    "#as in 2.5, but this time get the gene entries s.o. gene names\n",
    "for rec in BSIO.parse('geneEntry.gb', 'gb'):\n",
    "    for feature in rec.features:\n",
    "        if feature.type == \"gene\":\n",
    "            print (feature.qualifiers[\"gene\"][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.7\n",
    "\n",
    "Print the peptide sequence produced by the gene named \"BTK\" in the NCBI Nucleotide database entry 223468685.\n",
    "\n",
    "(Note that the CDS is discontinuous and consists of several regions, which is modelled by a `CompoundLocation` object.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BTK\n",
      "['MAAVILESIFLKRSQQKKKTSPLNFKKRLFLLTVHKLSYYEYDFERGRRGSKKGSIDVEKITCVETVVPEKNPPPERQIPRRGEESSEMEQISIIERFPYPFQVVYDEGPLYVFSPTEELRKRWIHQLKNVIRYNSDLVQKYHPCFWIDGQYLCCSQTAKNAMGCQILENRNGSLKPGSSHRKTKKPLPPTPEEDQILKKPLPPEPAAAPVSTSELKKVVALYDYMPMNANDLQLRKGDEYFILEESNLPWWRARDKNGQEGYIPSNYVTEAEDSIEMYEWYSKHMTRSQAEQLLKQEGKEGGFIVRDSSKAGKYTVSVFAKSTGDPQGVIRHYVVCSTPQSQYYLAEKHLFSTIPELINYHQHNSAGLISRLKYPVSQQNKNAPSTAGLGYGSWEIDPKDLTFLKELGTGQFGVVKYGKWRGQYDVAIKMIKEGSMSEDEFIEEAKVMMNLSHEKLVQLYGVCTKQRPIFIITEYMANGCLLNYLREMRHRFQTQQLLEMCKDVCEAMEYLESKQFLHRDLAARNCLVNDQGVVKVSDFGLSRYVLDDEYTSSVGSKFPVRWSPPEVLMYSKFSSKSDIWAFGVLMWEIYSLGKMPYERFTNSETAEHIAQGLRLYRPHLASEKVYTIMYSCWHEKADERPTFKILLSNILDVMDEES']\n",
      "\n",
      "\n",
      "This is BTKs peptide sequence in string format: \n",
      "MAAVILESIFLKRSQQKKKTSPLNFKKRLFLLTVHKLSYYEYDFERGRRGSKKGSIDVEKITCVETVVPEKNPPPERQIPRRGEESSEMEQISIIERFPYPFQVVYDEGPLYVFSPTEELRKRWIHQLKNVIRYNSDLVQKYHPCFWIDGQYLCCSQTAKNAMGCQILENRNGSLKPGSSHRKTKKPLPPTPEEDQILKKPLPPEPAAAPVSTSELKKVVALYDYMPMNANDLQLRKGDEYFILEESNLPWWRARDKNGQEGYIPSNYVTEAEDSIEMYEWYSKHMTRSQAEQLLKQEGKEGGFIVRDSSKAGKYTVSVFAKSTGDPQGVIRHYVVCSTPQSQYYLAEKHLFSTIPELINYHQHNSAGLISRLKYPVSQQNKNAPSTAGLGYGSWEIDPKDLTFLKELGTGQFGVVKYGKWRGQYDVAIKMIKEGSMSEDEFIEEAKVMMNLSHEKLVQLYGVCTKQRPIFIITEYMANGCLLNYLREMRHRFQTQQLLEMCKDVCEAMEYLESKQFLHRDLAARNCLVNDQGVVKVSDFGLSRYVLDDEYTSSVGSKFPVRWSPPEVLMYSKFSSKSDIWAFGVLMWEIYSLGKMPYERFTNSETAEHIAQGLRLYRPHLASEKVYTIMYSCWHEKADERPTFKILLSNILDVMDEES\n"
     ]
    }
   ],
   "source": [
    "# Make an SeqFeature? object\n",
    "for rec in BSIO.parse('geneEntry.gb', 'gb'):\n",
    "    # intialize variable feature to go through the features of the Seq object\n",
    "    for feature in rec.features:\n",
    "        #Determine that the we are looking for the word \"CDS\"\n",
    "        if feature.type == \"CDS\":\n",
    "            # initialize variable x to go through gene names under CDS\n",
    "            for x in feature.qualifiers.get(\"gene\"):\n",
    "                # the word is \"BTK\"\n",
    "                if x == \"BTK\":\n",
    "                    #then print the site_type x and the peptide seq\n",
    "                    print(x)\n",
    "                    print(feature.qualifiers[\"translation\"])\n",
    "                    print(\"\\n\")\n",
    "                    #just tried if I can convert this list to string\n",
    "                    print(\"This is \"+x+\"s\"+\" peptide sequence in string format: \\n\" + ' '.join([str(elem) for elem in feature.qualifiers[\"translation\"]]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2.8\n",
    "\n",
    "Fetch the sequences of those NCBI Protein database entries that concern the species Saccharomyces boulardii and save them to a file named `saccharomyces-boulardii-proteins.fasta` in the FASTA format. Use the query history and fetch the sequences in batches of 100 sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def retrieve(query, batch_size):\n",
    "    # search using history\n",
    "    handle = BE.esearch(db=\"protein\",\n",
    "                        term=query,\n",
    "                        usehistory=\"y\")\n",
    "    results = BE.read(handle)\n",
    "    handle.close()\n",
    "    webenv = results['WebEnv']\n",
    "    querykey = results['QueryKey']\n",
    " # adjust maximum size if less than maximum number of hits available\n",
    "   # retrieve_size = min(int(results['Count']), retrieve_size)\n",
    "    for i in range(0, batch_size):\n",
    "        handle = BE.efetch(db=\"protein\",\n",
    "            webenv=webenv,\n",
    "            query_key=querykey,\n",
    "                           # fetch N records starting from i\n",
    "            retstart=i,\n",
    "            retmax=batch_size,\n",
    "            rettype=\"fasta\",\n",
    "            retmode=\"text\")\n",
    "        for record in BSIO.parse(handle, 'fasta'):\n",
    "            yield record\n",
    "        handle.close()\n",
    "        \n",
    "for record in retrieve(\"Saccharomyces boulardii\", 100):\n",
    "    BSIO.write([record], 'saccharomyces-boulardii-proteins.fasta', 'fasta')\n",
    "   # print(record)\n",
    " \n",
    "#Didn't test if the file is readable as there are so many entries and it was said in the\n",
    "#wednesday session that there wasn't a need for printing\n",
    "\n",
    "\n",
    "#with open('saccharomyces-boulardii-proteins.fasta') as f:\n",
    "  #  print(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
